File: mixin-sleman/src/main/java/com/andia/mixin/sleman/util/FeatureCode.java
/*
 * Copyright (c) 2020-2023 Datamixin.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
package com.andia.mixin.sleman.util;

import static java.lang.Integer.toHexString;

import java.util.Arrays;
import java.util.List;

import com.andia.mixin.model.EFeature;
import com.andia.mixin.model.EList;
import com.andia.mixin.model.EObject;
import com.andia.mixin.sleman.api.SExpression;
import com.andia.mixin.sleman.model.XExpression;

public class FeatureCode {

	private XExpression expression;

	public FeatureCode(SExpression expression) {
		this.expression = (XExpression) expression;
	}

	public String generate() {

		EObject current = expression;
		StringBuffer buffer = new StringBuffer();
		while (current != null) {

			EFeature feature = current.eContainingFeature();
			EObject container = current.eContainer();
			if (container != null) {

				Object value = container.eGet(feature);
				if (value instanceof EList) {
					EList<?> list = (EList<?>) value;
					encodeElement(buffer, list, current, true);
				}

				EFeature[] features = container.eFeatures();
				List<EFeature> list = Arrays.asList(features);
				encodeElement(buffer, list, feature, false);
			}
			current = container;
		}
		return buffer.toString();
	}

	private void encodeElement(StringBuffer buffer, List<?> list, EFeature element, boolean flag) {
		int index = list.indexOf(element);
		encodeIndex(buffer, index, flag);
	}

	private void encodeElement(StringBuffer buffer, EList<?> list, Object element, boolean flag) {
		int index = list.indexOf(element);
		encodeIndex(buffer, index, flag);
	}

	private void encodeIndex(StringBuffer buffer, int index, boolean flag) {
		String hex = toHexString(index);
		buffer.insert(0, hex);
		if (flag) {
			int length = hex.length();
			String lengthHex = toHexString(length);
			buffer.insert(0, lengthHex);
		}
	}

}

File: mixin-sleman-webface/src/main/js/sleman/util/FeatureCode.ts
/*
 * Copyright (c) 2020-2023 Datamixin.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
import EList from "webface/model/EList";
import EObject from "webface/model/EObject";

export default class FeatureCode {

    private model: EObject;

    constructor(model: EObject) {
        this.model = model;
    }

    public generate(): string {

        let current: EObject = this.model;
        let buffer: string[] = [];
        while (current != null) {

            let feature = current.eContainingFeature();
            let container = current.eContainer();
            if (container != null) {

                let value = container.eGet(feature);
                if (value instanceof EList) {
                    let list = value.toArray();
                    this.encodeElement(buffer, list, current, true);
                }

                let features = container.eFeatures();
                this.encodeElement(buffer, features, feature, false);
            }
            current = container;
        }
        return buffer.join("");
    }

    private encodeElement(buffer: string[], list: any[], element: any, flag: boolean): void {
        let index = list.indexOf(element);
        this.encodeIndex(buffer, index, flag);
    }

    private encodeIndex(buffer: string[], index: number, flag: boolean): void {
        let hex = index.toString(16);
        buffer.splice(0, 0, hex);
        if (flag) {
            let length = hex.length;
            let lengthHex = length.toString(16);
            buffer.splice(0, 0, lengthHex);
        }
    }

    public static generate(model: EObject): string {
        let code = new FeatureCode(model);
        return code.generate();
    }

}
